### DISCUSSION TITLE
Overwrite frozen main.py

### DISCUSSION THREAD

**Discussion by rolfbachmann** [micropython/micropython#15254] [Using MicroPython]
Hello, 

I've a RP2 project where I build my own firmware with some frozen python files ([with manifest](https://docs.micropython.org/en/latest/reference/manifest.html)), say `version.py` and `main.py`. Works just fine, device boots up and run the `main.py`.

Now I need to do some further development, changed my `main.py` file and upload it to the device (by using `mpremote`), but the changes will not be executed. If I change the version number in `version.py`, which will be imported by `main.py` and upload the `version.py` this seems to work, I get another version number.

I found a comment from @robert-hh (https://github.com/orgs/micropython/discussions/14144#discussioncomment-8862144) where he said "there is a hiccup with `main.py`". I don't know if it's still valid.

My question:

- What's the best workflow to do further development on the frozen code?
- Does it make sense to move as much code as possible to other modules so I don't need to change the `main.py`?

I searched around for a while, but didn't find the right answer. Maybe I used the wrong search terms.

**Comment by andrewleech** [comment]
As long as you haven't also frozen boot.py, you should be able to use that to change the setup to load your new main

First thing to try is to change the order of `sys.path`. This list controls import location priority, where `.frozen` represents the frozen in code. If you insert `/` or `/flash` at the start of the list (can't remember which is appropriate for rp2) then it should get used before frozen.

With regard to the link, I know some ports support frozen main, but perhaps some others don't still? I'm not sure here, the `main.c` which handles the internal startup logic is separate for each port (has to handle custom peripherals etc) so there's some discrepancy here still.

**Reply by rolfbachmann** [reply]
Thank you for your reply. Gave me some idea to try out.

The `sys.path` (`['', '.frozen', '/lib']`) works as expected. If I import a module from the repl it uses the "new" uploaded, not the "frozen" module. And if I do `import main` the new `main` module will be loaded and executed.

It only don't work if reset the device (e.g. soft reset). Then the `main` from "frozen" will be used.

I think this is this hiccup described in the referred comment.

I'll move the most code out of `main` so it don't need to be updated regularly. And if I've to, I'll build a new firmware.

**Reply by robert-hh** [reply]
It looks that I was wrong in my comment cited above. Like you noticed a frozen main.py will be executed if it exists. That was changed in 2019 (oops!).

**Reply by davefes** [reply]
Just a suggestion if CRTL-C is a problem.  Maybe, start main.py off with:
```py
# during development
if machine.reset_cause() == machine.SOFT_RESET:
    print('you got a SOFT_RESET, so do a machine.reset()')
    time.sleep(1)
    machine.reset()
```
and then some code in main.py if you want.

**Reply by rolfbachmann** [reply]
I've tried soft and hard reset with the same result: Reset always use the frozen `main.py`. I've to `import main` to run the "new" main. But it's ok for my dev workflow, if I really need to change the `main.py`.

**Reply by peterhinch** [reply]
One option is not to freeze `main.py` - put it in the filesystem. It's fine to mix frozen and non-frozen modules. Leave modules that are subject to change in the filesystem.

**Reply by bikeNomad** [reply]
I've done this in my frozen `main.py` to allow for `/main.py` to override:
```
import os

# allow for filesystem /main.py to override this one
root_files = os.listdir()
if 'main.py' in root_files:
    import main

else:
   # do normal stuff
```
And you can do the same thing in the frozen `boot.py` (after setting up `sys.path`):
```
root_files = os.listdir()
if 'boot.py' in root_files:
    import boot
```
Interestingly, what I get when I do both is this:
```
MPY: soft reboot
.frozen/boot: sys.path=['', '.frozen/app', '.frozen']
this is FS /boot.py
This is FS /main.py
start frozen main.py
```
That is, the `.frozen/main` runs _after_ `/main.py`.
